backends/bmv2/simple_switch/simpleSwitch.cpp

```diff
@@ -0,0 +1,970 @@
+    structure->pipeline_controls.emplace(egress_name);
+
+    auto compute = main->findParameterValue(v1model.sw.compute.name);
+    if (compute == nullptr || !compute->is<IR::ControlBlock>()) {
+        modelError("%1%: main package does not match the expected model", main);
+        return false;
+    }
+    structure->block_type.emplace(compute->to<IR::ControlBlock>()->container, V1_COMPUTE);
+    structure->non_pipeline_controls.emplace(compute->to<IR::ControlBlock>()->container->name);
+
+    auto deparser = main->findParameterValue(v1model.sw.deparser.name);
+    if (deparser == nullptr || !deparser->is<IR::ControlBlock>()) {
+        modelError("%1%: main package  match the expected model", main);
+        return false;
+    }
+    structure->block_type.emplace(deparser->to<IR::ControlBlock>()->container, V1_DEPARSER);
+    structure->non_pipeline_controls.emplace(deparser->to<IR::ControlBlock>()->container->name);
+
+    return false;
+}
+
+EXTERN_CONVERTER_SINGLETON(clone)
+EXTERN_CONVERTER_SINGLETON(clone3)
+EXTERN_CONVERTER_SINGLETON(hash)
+EXTERN_CONVERTER_SINGLETON(digest)
+EXTERN_CONVERTER_SINGLETON(resubmit)
+EXTERN_CONVERTER_SINGLETON(recirculate)
+EXTERN_CONVERTER_SINGLETON(mark_to_drop)
+EXTERN_CONVERTER_SINGLETON(random)
+EXTERN_CONVERTER_SINGLETON(truncate)
+EXTERN_CONVERTER_SINGLETON(register)
+EXTERN_CONVERTER_SINGLETON(counter)
+EXTERN_CONVERTER_SINGLETON(meter)
+EXTERN_CONVERTER_SINGLETON(direct_counter)
+EXTERN_CONVERTER_SINGLETON(direct_meter)
+EXTERN_CONVERTER_SINGLETON(action_profile)
+EXTERN_CONVERTER_SINGLETON(action_selector)
+
+CONVERT_EXTERN_FUNCTION(clone) {
+    int id = -1;
+    if (mc->arguments->size() != 2) {
+        modelError("Expected 2 arguments for %1%", mc);
+        return nullptr;
+    }
+    cstring name = ctxt->refMap->newName("fl");
+    auto emptylist = new IR::ListExpression({});
+    id = createFieldList(ctxt, emptylist, "field_lists", name, ctxt->json->field_lists);
+
+    auto cloneType = mc->arguments->at(0);
+    auto ei = P4::EnumInstance::resolve(cloneType->expression, ctxt->typeMap);
+    if (ei == nullptr) {
+        modelError("%1%: must be a constant on this target", cloneType);
+        return nullptr;
+    }
+    cstring prim = ei->name == "I2E" ? "clone_ingress_pkt_to_egress" :
+                   "clone_egress_pkt_to_egress";
+    auto session = ctxt->conv->convert(mc->arguments->at(1)->expression);
+    auto primitive = mkPrimitive(prim);
+    auto parameters = mkParameters(primitive);
+    // TODO(jafingerhut):
+    // primitive->emplace_non_null("source_info", s->sourceInfoJsonObj());
+    parameters->append(session);
+
+    if (id >= 0) {
+        auto cst = new IR::Constant(id);
+        ctxt->typeMap->setType(cst, IR::Type_Bits::get(32));
+        auto jcst = ctxt->conv->convert(cst);
+        parameters->append(jcst);
+    }
+    return primitive;
+}
+
+CONVERT_EXTERN_FUNCTION(clone3) {
```

Comment:
I expect you may be able to reuse some of this code for psa as well. Maybe it will need to move in a base class.